monads/src/HW/Eval.hs

```diff
@@ -20,17 +20,54 @@ data MachineState v = MachineState
   { getStack :: Stack,
     getEnv :: Env v
   }
-  deriving (Show)
+  deriving (Show, Eq)
 
 -- Run the compiled program on an empty stack and environment
 initialState :: MachineState String
 initialState = MachineState [] M.empty
 
 -- Execute a single instruction. 
 -- Successful evaluation does not produce any useful result: only the effect of modifying state matters. 
+
 execInstr :: (Ord v, Show v) => StackInstr v -> MyState (MachineState v) (Either (Error v) ())
-execInstr = undefined 
+execInstr (PushNum value) = do 
+  modify (\s -> s {getStack = value : getStack s})
+  return $ Right ()
+
+execInstr (PushVar name) = do
+  map <- gets getEnv
+  case M.lookup name map of
+    Just x -> do
+      modify (\s -> s {getStack = x : getStack s})
+      return $ Right ()
```

Comment:
avoid `return`

Reply:
Hi, I would need a clarification for this, do requires return statement or the fully constructed MyState, I provided two other ways but they are not better than the one that I provided, avoiding return in this case is useless for me, it makes me write more code and is not as readable as before. Could you please clarify why do I need to avoid the return?

Reply:
Use `fmap`.

Reply:
Fixed